### PR TITLE
Use associated types for layout wrapper trait, and generalized ThreadSafeLayoutFoo

### PR DESCRIPTION
**Pull request by bholley**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8639)

### CODE REVIEW

**Comment by bholley** [comment]
This is a continuation of the work in #8595.

File: components/layout/wrapper.rs [hunk 1]
```diff
@@ -782,35 +792,203 @@ impl<T> PseudoElementType<T> {
...
+
+    /// If this node is an iframe element, returns its pipeline ID. If this node is
+    /// not an iframe element, fails.
+    fn iframe_pipeline_id(&self) -> PipelineId;
+
+    fn get_colspan(&self) -> u32;
+}
+
+// These can violate the thread-safety and therefore are not public.
+trait DangerousThreadSafeLayoutNode<'ln> : ThreadSafeLayoutNode<'ln> {
```
Comment (pcwalton) [inline components/layout/wrapper.rs]: So as I typed this I realized why you didn't use `UnsafeThreadSafeLayoutNode`, as it's a bit of an oxymoron of a name :) But I still lean toward consistency in use of "Unsafe". How about `MemoryUnsafeThreadSafeLayoutNode`, to maintain consistency in naming without the oxymoron?

Comment (bholley) [inline components/layout/wrapper.rs]: I could, but we already have something called UnsafeLayoutNode, which would be super confusing IMO - that's why I opted for a different name. :-)

**Comment by bholley** [comment]
Since you're looking at it... @pcwalton r?

File: components/layout/wrapper.rs (hunk 1 above)
Comment (bholley) [inline components/layout/wrapper.rs]: And in general, this name doesn't escape anywhere and is used in only a few places in this file, so I think having it be consistent is less important.

**Comment by pcwalton** [comment]
Sure, seems fine.

**Pull request merged**